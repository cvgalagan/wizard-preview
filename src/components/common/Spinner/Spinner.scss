@import "../../../styles/common";

.spinner {
    display: inline-block;
    $self: &;
    &_indeterminate {
        animation: CircularRotate 1.4s linear infinite;
        #{$self}__circle {
            animation: CircularDash 1.4s ease-in-out infinite;
            stroke-dasharray: 80px, 200px;
            stroke-dashoffset: 0;
        }
    }
    &_static {
        transition: transform .2s ease-in-out;
        #{$self}__circle {
            transition: stroke-dashoffset .2s ease-in-out;
        }
    }
    &_disable-shrink {
        #{$self}__circle {
            animation: none;
        }
    }
    &__svg {
        display: block;
    }
    &__circle {
        width: 100%;
        height: 100%;
        stroke: currentColor;
    }
}

@keyframes CircularRotate {
    100% {
        transform: rotate(360deg);
    }
}

@keyframes CircularDash {
    0% {
        stroke-dasharray: 1px, 200px;
        stroke-dashoffset: 0;
    }
    50% {
        stroke-dasharray: 100px, 200px;
        stroke-dashoffset: -15px;
    }
    100% {
        stroke-dasharray: 100px, 200px;
        stroke-dashoffset: -125px;
    }
}