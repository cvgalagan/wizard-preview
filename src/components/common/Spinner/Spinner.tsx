import React, { CSSProperties } from "react"
import "./Spinner.scss"
import cn from "classnames"

const SIZE = 44

export interface SpinnerProps extends React.HTMLAttributes<HTMLDivElement> {
    disableShrink?: boolean
    size?: number | string
    thickness?: number
    value?: number
    variant?: "determinate" | "indeterminate" | "static"
}

function getRelativeValue(value: number, min: number, max: number) {
    return (Math.min(Math.max(min, value), max) - min) / (max - min)
}

function easeOut(t: number) {
    t = getRelativeValue(t, 0, 1)
    t = (t -= 1) * t * t + 1
    return t
}

function easeIn(t: number) {
    return t * t
}

const Spinner = React.forwardRef<HTMLDivElement, SpinnerProps>(function (props, ref) {
    const {
        className,
        disableShrink = false,
        size,
        thickness = 3.6,
        value = 0,
        variant = "indeterminate",
        ...other
    } = props

    const circleStyle: CSSProperties = {}
    const rootStyle: CSSProperties = {}
    const rootProps: React.HTMLAttributes<HTMLDivElement> = {}

    if (variant === "determinate" || variant === "static") {
        const circumference = 2 * Math.PI * ((SIZE - thickness) / 2)
        circleStyle.strokeDasharray = circumference.toFixed(3)
        rootProps["aria-valuenow"] = Math.round(value)

        if (variant === "static") {
            circleStyle.strokeDashoffset = `${(((100 - value) / 100) * circumference).toFixed(3)}px`
            rootStyle.transform = "rotate(-90deg)"
        } else {
            circleStyle.strokeDashoffset = `${(easeIn((100 - value) / 100) * circumference).toFixed(3)}px`
            rootStyle.transform = `rotate(${(easeOut(value / 70) * 270).toFixed(3)}deg)`
        }
    }

    return (
        <div
            className={cn(
                "spinner",
                variant === "indeterminate" ? "spinner_indeterminate" : variant === "static" && "spinner_static",
                disableShrink && "spinner_disable-shrink",
                className
            )}
            style={{ width: size, height: size, ...rootStyle }}
            ref={ref}
            role="progressbar"
            {...rootProps}
            {...other}
        >
            <svg className={"spinner__svg"} viewBox={`${SIZE / 2} ${SIZE / 2} ${SIZE} ${SIZE}`}>
                <circle
                    className="spinner__circle"
                    style={circleStyle}
                    cx={SIZE}
                    cy={SIZE}
                    r={(SIZE - thickness) / 2}
                    fill="none"
                    strokeWidth={thickness}
                />
            </svg>
        </div>
    )
})

export default Spinner
