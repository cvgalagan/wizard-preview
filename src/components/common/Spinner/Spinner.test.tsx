import React from "react"
import { shallow } from "enzyme"
import Spinner from "./Spinner"

describe("Spinner", () => {
    it("should render", () => {
        const wrapper = shallow(<Spinner />)
        expect(wrapper).toMatchSnapshot()
    })
    it("should add 'indeterminate' modifier by default", () => {
        const wrapper = shallow(<Spinner />)
        const element = wrapper.find(".spinner").first()
        expect(element.hasClass("spinner_indeterminate")).toEqual(true)
    })
    it("should add 'static' modifier if variant is indeterminate", () => {
        const wrapper = shallow(<Spinner variant="indeterminate" />)
        const element = wrapper.find(".spinner").first()
        expect(element.hasClass("spinner_indeterminate")).toEqual(true)
    })
    it("should add 'static' modifier if variant is static", () => {
        const wrapper = shallow(<Spinner variant="static" />)
        const element = wrapper.find(".spinner").first()
        expect(element.hasClass("spinner_static")).toEqual(true)
    })
    it("should add 'disable-shrink' modifier if disableShrink is true", () => {
        const wrapper = shallow(<Spinner disableShrink />)
        const element = wrapper.find(".spinner").first()
        expect(element.hasClass("spinner_disable-shrink")).toEqual(true)
    })
})
