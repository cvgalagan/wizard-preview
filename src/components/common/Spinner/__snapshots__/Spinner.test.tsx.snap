// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner should render 1`] = `
<div
  className="spinner spinner_indeterminate"
  role="progressbar"
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <svg
    className="spinner__svg"
    viewBox="22 22 44 44"
  >
    <circle
      className="spinner__circle"
      cx={44}
      cy={44}
      fill="none"
      r={20.2}
      strokeWidth={3.6}
      style={Object {}}
    />
  </svg>
</div>
`;
